File: spring-boot-app/src/main/java/org/example/angulardemo/controller/ProductController.java
package org.example.angulardemo.controller;

import jakarta.validation.Valid;
import org.example.angulardemo.entity.Product;
import org.example.angulardemo.repo.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;

@RestController
@CrossOrigin(origins = "http://localhost")
@RequestMapping("/products")
public class ProductController {
    private final ProductRepository productRepository;

    @Autowired
    public ProductController(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    @GetMapping
    public Iterable<Product> getAllProducts() {
        return productRepository.findAll();
    }

    @PostMapping
    public ResponseEntity<?> addProduct(@Valid @RequestBody Product newProduct, BindingResult bindingResult) {
        if (bindingResult.hasErrors()) {
            return new ResponseEntity<>(bindingResult.getAllErrors(), HttpStatus.BAD_REQUEST);
        }

        if (newProduct.getId() != null) {
            return new ResponseEntity<>("Id field should not be set", HttpStatus.BAD_REQUEST);
        }

        return new ResponseEntity<>(productRepository.save(newProduct), HttpStatus.CREATED);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteProduct(@PathVariable Long id) {
        return productRepository.deleteProductById(id) > 0 ? new ResponseEntity<>(HttpStatus.NO_CONTENT) : new ResponseEntity<>(HttpStatus.NOT_FOUND);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateProduct(@PathVariable Long id, @Valid @RequestBody Product productUpdate, BindingResult bindingResult) {
        if (bindingResult.hasErrors()) {
            return new ResponseEntity<>(bindingResult.getAllErrors(), HttpStatus.BAD_REQUEST);
        }
        return productRepository.updateProductById(id, productUpdate.getExtId(), productUpdate.getName(), productUpdate.getDescription()) > 0
                ? new ResponseEntity<>(HttpStatus.OK) : new ResponseEntity<>(HttpStatus.NOT_FOUND);
    }
}

File: spring-boot-app/src/main/kotlin/org/example/angulardemo/controller/ProductController.kt
package org.example.angulardemo.controller

import jakarta.validation.Valid
import kotlinx.coroutines.flow.Flow
import org.example.angulardemo.dto.ProductDTO
import org.example.angulardemo.service.ProductService
import org.springframework.http.HttpStatus
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*

@RestController
@Validated
@RequestMapping("/products")
class ProductController(
    val productService: ProductService,
) {

    @PostMapping("{userId}")
    @ResponseStatus(HttpStatus.CREATED)
    suspend fun addProduct(
        @PathVariable("userId") userId: Long,
        @RequestBody @Valid productDTO: ProductDTO,
    ): ProductDTO =
        productService.addProduct(userId, productDTO)

    @GetMapping("/{userId}")
    suspend fun getAllProducts(
        @PathVariable("userId") userId: Long,
    ): Flow<ProductDTO> = productService.getAllProductsForUser(userId)

    @PutMapping("/{userId}/{id}")
    suspend fun updateProduct(
        @PathVariable("userId") userId: Long,
        @PathVariable("id") productId: Long,
        @RequestBody @Valid productDTO: ProductDTO,
    ): ProductDTO =
        productService.updateProduct(userId, productId, productDTO)

    @DeleteMapping("/{userId}/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    suspend fun deleteCourse(
        @PathVariable("userId") userId: Long,
        @PathVariable("id") productId: Long,
    ) = productService.deleteProduct(userId, productId)

}
